File: Data Structures & Algorithms/palindromic-substrings/submission-0.java
class Solution {
    public int countSubstrings(String s) {
        int res = 0;
        int n = s.length();
        boolean [][] dp = new boolean[n][n];

        for(int i = 0; i<n; i++) {
            for(int j = 0; j<=i; j++) {
                if(s.charAt(i) == s.charAt(j) &&
                (i-j <= 2 || dp[i-1][j+1])) {
                    dp[i][j] = true;
                    res += 1;
                }
            }
        }

        return res;
    }
}

File: Data Structures & Algorithms/subsets-ii/submission-0.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> sub;

        sort(nums.begin(), nums.end());
        backtrack(nums, res, sub, 0);
        return res;
    }

    void backtrack(vector<int>& nums, vector<vector<int>>& res, vector<int>& sub, int i) {
        if(i == nums.size()) {
            res.push_back(sub);
            return;
        }

        // take
        sub.push_back(nums[i]);
        backtrack(nums, res, sub, i+1);

        // dont take (or already analyzed)
        sub.pop_back();
        while(i + 1 < nums.size() && nums[i] == nums[i+1]) {
            i++;
        }
        backtrack(nums, res, sub, i+1);
    }
};
